pkg/workloadmeta/collectors/containerd/event_builder.go

```diff
@@ -11,42 +11,68 @@ import (
 	"context"
 	"fmt"
 
+	"github.com/containerd/containerd/api/events"
 	containerdevents "github.com/containerd/containerd/events"
+	"github.com/gogo/protobuf/proto"
 
 	cutil "github.com/DataDog/datadog-agent/pkg/util/containerd"
+	"github.com/DataDog/datadog-agent/pkg/util/log"
 	"github.com/DataDog/datadog-agent/pkg/workloadmeta"
 )
 
 // buildCollectorEvent generates a CollectorEvent from a containerdevents.Envelope
-func buildCollectorEvent(ctx context.Context, containerdEvent *containerdevents.Envelope, containerdClient cutil.ContainerdItf) (workloadmeta.CollectorEvent, error) {
+func (c *collector) buildCollectorEvent(ctx context.Context, containerdEvent *containerdevents.Envelope) (workloadmeta.CollectorEvent, error) {
 	switch containerdEvent.Topic {
 	case containerCreationTopic, containerUpdateTopic:
 		ID, hasID := containerdEvent.Field([]string{"event", "id"})
 		if !hasID {
 			return workloadmeta.CollectorEvent{}, fmt.Errorf("missing ID in containerd event")
 		}
 
-		return createSetEvent(ctx, ID, containerdClient)
+		return createSetEvent(ctx, ID, c.containerdClient)
 	case containerDeletionTopic:
 		ID, hasID := containerdEvent.Field([]string{"event", "id"})
 		if !hasID {
 			return workloadmeta.CollectorEvent{}, fmt.Errorf("missing ID in containerd event")
 		}
 
-		return createDeletionEvent(ID), nil
-	case TaskStartTopic, TaskOOMTopic, TaskExitTopic, TaskDeleteTopic, TaskPausedTopic, TaskResumedTopic:
-		// Notice that the ID field in this case is stored in "ContainerID".
-		ID, hasID := containerdEvent.Field([]string{"event", "container_id"})
-		if !hasID {
-			return workloadmeta.CollectorEvent{}, fmt.Errorf("missing ID in containerd event")
+		exitInfo := c.getExitInfo(ID)
+		defer c.deleteExitInfo(ID)
+
+		return createDeletionEvent(ID, exitInfo), nil
+	case TaskExitTopic:
+		exited := &events.TaskExit{}
+		if err := proto.Unmarshal(containerdEvent.Event.Value, exited); err != nil {
+			log.Debugf("Could not unmarshal task exit event: %w", err)
+		}
+
+		c.cacheExitInfo(exited.ContainerID, &exited.ExitStatus, exited.ExitedAt)
```

Comment:
Shouldn't we do this only when `err` is nil?
Same for the `case` below this one.